### ISSUE TITLE
Cut / Copy Not Working in iPad OS

### ISSUE THREAD

**Issue by anthonygualandri** [microsoft/vscode#142057]
Looks like it's all working for me now.

_Originally posted by @TylerLeonhardt in https://github.com/microsoft/vscode/issues/101564#issuecomment-906629795_

You all seem to say this is working for you now, but I’m still getting the same problem as the issue above describes. What was the fix or is this actually still ongoing?

**Comment by TylerLeonhardt** [comment]
can you share if you're using vscode.dev or insiders.vscode.dev and what version of iPadOS you're using?

**Comment by anthonygualandri** [comment]
Mmmmm, I'm using github.dev (vs code inside of Codespaces) so I'm wondering if that means that this hasn't been fixed within the codespaces implementation side of things. I'm using the most current version of iPadOS, so 15.3. Thanks!

**Comment by TylerLeonhardt** [comment]
@anthonygualandri have you tried this today? We just released a new stable version of VS Code which I believe should have a fix for this. Ensure Codespaces is using version 1.64 of VS Code.

**Comment by anthonygualandri** [comment]
@TylerLeonhardt thanks for the reply! So it looks like the version used when I fire up the codespace is 1.64:

Version: 1.64.0
Commit: 5554b12acf27056905806867f251c859323ff7e9
Date: 2022-02-03T04:19:17.872Z
Browser: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.3 Safari/605.1.15

Any other things it could be?

**Comment by TylerLeonhardt** [comment]
Thanks for trying. It seems to be hit or miss because @rebornix and I can't repro... but someone else on our team has. @anthonygualandri can you try copy/paste in here:
https://microsoft.github.io/monaco-editor/

this will help us narrow down where the problem may be.

**Comment by anthonygualandri** [comment]
No problem glad to help! I just bought a brand new iPad thinking I was going to be able to use it as a laptop replacement as I do all def stuff in codespaces. So not having copy and paste definitely won’t work hahaha. I tried using that page you linked to and I can copy, cut and paste just fine with the on screen apple keyboard the vs code menu when long pressing the screen and the external keyboard. However, I was only able to under and redo with the external keyboard for some reason.

**Comment by anthonygualandri** [comment]
@TylerLeonhardt Just checking back to see if any updates on finding the sources of the issues with this. Thanks!

**Comment by anthonygualandri** [comment]
I tried copying and pasting again today this time I got a message from vs code in the browser saying that clipboard is being blocked by the browser with "Learn More" button that took me to the following url: 

https://code.visualstudio.com/docs/remote/codespaces#_how-do-i-allow-vs-code-to-access-my-clipboard-for-reading

I couldn't find a way to get to this setting on the ipad os browsers though, either Safari or Chrome. Wondering if you all ran into this at all and know a way to enable, or if this is actually not the problem? Thanks!

**Comment by TylerLeonhardt** [comment]
Sorry for the delay. Spinning many plates 😅  

> I couldn't find a way to get to this setting on the ipad os browsers though

This dialog doesn't show up on iPad. The browser just doesn't do it, for some reason. It only seems to happen on Desktop Chrome and Chromium based browsers.

With that said..... let me summarize here so I can formulate what to do next:

Cut/Copy doesn't work for you at all if you use either an external keyboard or the on screen keyboard
You reproduce this in GitHub Codespaces
You are unable to repro this behavior in https://microsoft.github.io/monaco-editor/

Were you able to repro this using https://vscode.dev ?

**Comment by anthonygualandri** [comment]
No problem. Totally understand. Glad to hear you guys have a fix in mind that you can look to implement next week. So in terms of trying this in vscode.dev, I just tried it now and cut, copy and paste work with both direct iPad long pressing as well as external keyboard. Same as with the repo env link you sent me to test, I can undo and redo with the external keyboard, but the on screen keyboard does not do undo and redo. I see the back and forward arrows display as usable (unlike in the codespace.do env), but when I press them they don't work. Anyway, hopefully that helps further pin point things for you. 

What are the differences between vscode.dev and codespaces.dev? Does vscode.dev keep all local package installation, etc, remote like codespaces.dev? With vscode.dev can you install extensions and debug locally like with codespaces.dev?

Thanks for all the help!

**Comment by TylerLeonhardt** [comment]
I would have expected that vscode.dev & codespaces would behave the same... but I guess not. @jkeech @joaomoreno any ideas why there might be a difference in behavior here?

**Comment by jkeech** [comment]
@anthonygualandri, can you also try on github.dev using the same repository that you try on vscode.dev? There are three different websites that are similar but distinct. Knowing which one(s) are working will help narrow it down:

1) vscode.dev (the lightweight editor in the browser) - example: https://vscode.dev/github/microsoft/vscode
2) github.dev (the lightweight editor in the browser) - example: https://github.dev/microsoft/vscode
3) Codespaces (*.github.dev) - Create a new codespace from `microsoft/vscode`, it will give you `some-random-name.github.dev`

(1) and (2) are both completely client side in the browser. (3) is partially client side and partially running on the remote container. If it's working for you in (1), I'd also expect it to work in (2) since those are very similar to each other. If it doesn't work in (2), that would be interesting to know. Even for (3), it should be similar enough to (1) and (2) to not make a difference here, but I'm not as familiar with the implementation details between the pure client side version and the remote-attached version.

**Comment by anthonygualandri** [comment]
@jkeech Ok, so I tried things in the GitHub.dev env and it works, as expected there as with vscode.dev. I can copy, paste and cut with on screen keyboard and external. I can also undo and redo with the keyboard and with the fly out menu inside the code editor, but not with the on screen iPad keyboard. For the codespaces env, I wasn't sure how to create a new codespace from microsoft/vscode as you can only select a repo that's currently yours. I created a new one from an existing repo I have and the behavior there is the same as with my original codespace where I noticed the problem initially. No copy, paste, cut, redo, undo of any kind. I'm currently importing the microsoft/vscode repo into my account from GitHub, but it's taking forever to complete (probably because it's so huge!), but I expect the behavior to be consistent with the experiences till now. So think you know where to target for a fix with this?

**Comment by jkeech** [comment]
> so I tried things in the GitHub.dev env and it works, as expected there as with vscode.dev.

Thanks for trying that out! It looks like this issue is something that only happens when running VS Code in the web connected to a remote backend (like Codespaces). @TylerLeonhardt, is there something different about the way they keybindings work in a remote-attached scenario compared to a serverless/client-side scenario? Codespaces isn't initializing the VS Code workbench differently in the two flows, so I don't think this is something that the embedder is doing to interfere with the cut/copy/paste on iPad.

**Comment by TylerLeonhardt** [comment]
So I was able to repro this in Codespaces. Here’s something very interesting:

Codespaces:
![B8A34E3A-73BE-480D-8763-EC427EE499D8](https://user-images.githubusercontent.com/2644648/156436637-890533a4-2aa8-4e9e-be6c-7b39c6fb6945.jpeg)
a web connected remote backend that isn’t Codespaces:
![B1C79A0F-A045-4FC1-8856-0DE02B3F80D1](https://user-images.githubusercontent.com/2644648/156436641-1f9e6e20-915d-411e-8b07-bb29e632b159.jpeg)

The cursor appears to be different… however, tbh I don’t know what this means exactly… all I know is that with the circle cursor, copy works… without it, it doesn’t.

CC @alexdima curious if you have any ideas

Oh yeah, I did notice that copying in the search view or the find widget or the terminal… all of that **worked**. It was only the _editor_ that had an issue. To attempt to help with this I added the commit above which forces cmd+c to go through our copy code (instead of relying on the browser to do the right thing).

We have [a wild workaround in place](https://github.com/microsoft/vscode/blob/6b9dde1cd3686e1b95b90388a0995187c1c63020/src/vs/platform/clipboard/browser/clipboardService.ts#L44) only for Safari that allows behaviors like copying from the terminal to work correctly and @rebornix and I’s suspicion is that that is causing copy in the editor to fail sometimes.

**Comment by alexdima** [comment]
@TylerLeonhardt How's the editor at  https://microsoft.github.io/monaco-editor/ behaving? (That would rule out the wild workaround).

**Comment by TylerLeonhardt** [comment]
@alexdima yeah that link works totally fine. Same for @anthonygualandri (at summarized here https://github.com/microsoft/vscode/issues/142057#issuecomment-1050409696 )

**Comment by anthonygualandri** [comment]
> The cursor appears to be different… however, tbh I don’t know what this means exactly… all I know is that with the circle cursor, copy works… without it, it doesn’t.

I think the circle cursor indicates the iPad gesture way of selecting text using (ie using your fingures or a stylus vs using a keyboard). Apple has this hybrid mouse / gesture thing going which is mucking everything up for sure. When I use the "mouse" on the iPad connected keyboard and I right click it shows me the gestures looking copy, paste, etc buttons like as if I was on my phone or iPad without a keyboard connected and doing it with my finger. But if I highlight some text on the screen with the "mouse" using the keyboard and then cmd+c to copy it, I don't get that circle cursor thing or the bubble copy/paste buttons. It just copies the text. So basically, this is Apple forcing you to program things on your end to handle their frankenstein way of having given mouse/keyboard functionality to the iPad, instead of doing it the computer laptop way the rest of the software world is used to.

I'll check back Friday to test out using Insiders. Thanks @TylerLeonhardt for working on this!

**Comment by TylerLeonhardt** [comment]
@alexdima also to summarize, I was able to repro in Codespaces... I remember Kai had mentioned similar copy issues in vscode.dev, though I wasn't able to repro... and @rebornix wasn't able to repro in Codespaces OR vscode.dev.

My theory is that there could be a race condition [between these handlers in our wild workaround](https://github.com/microsoft/vscode/blob/6b9dde1cd3686e1b95b90388a0995187c1c63020/src/vs/platform/clipboard/browser/clipboardService.ts#L68-L69) and the browser's handler of Copy natively and that my change essentially forces cmd+c to always use our service... but that's just a hunch.

**Comment by anthonygualandri** [comment]
Not sure if you changed anything yet, but I'm still getting the same behavior.

**Comment by anthonygualandri** [comment]
@TylerLeonhardt Swinging back to this to see if there's any new updates since the last little burst back and forth. Hope all's well!

**Comment by alexdima** [comment]
We might have repro steps for this on macos via https://github.com/microsoft/vscode/issues/148154

**Comment by alexdima** [comment]
From investigating https://github.com/microsoft/vscode/issues/148154, I discovered that `document.execCommand('cut')` and `document.execCommand('copy')` are no-ops in Safari when the `<textarea>` is empty (i.e. nothing selected). I think that perhaps something similar might be going on in the iPad. I pushed  https://github.com/microsoft/vscode/commit/155389030c0741ef9dc76da7b03b67559d34ee74 to always have something selected in the `<textarea>` when there is a selection in the editor. I will optimistically close this issue too, because I think the change will help the iPad case too.
